principal/redisproxy/redisproxy.go

```diff
@@ -0,0 +1,806 @@
+}
+
+// argoCDRedisWriterInternal is a simple wrapper around argo cd redis socket, which ensures that lock is owned while writing to argo cd server.
+type argoCDRedisWriterInternal struct {
+
+	// Use 'writeToArgoCDRedisSocket' function: these internal fields should not be called directly.
+
+	fromArgoCDWrite *bufio.Writer
+	argoCDWriteLock *sync.Mutex
+}
+
+// writeToArgoCDRedisSocket writes data back to Argo CD while owning mutex
+func (are *argoCDRedisWriterInternal) writeToArgoCDRedisSocket(logCtx *logrus.Entry, bytes []byte) error {
+
+	are.argoCDWriteLock.Lock()
+	defer are.argoCDWriteLock.Unlock()
+
+	_, err := are.fromArgoCDWrite.Write(bytes)
+	if err != nil {
+		log().WithError(err).Error("unable to write to Argo CD Redis socket")
+		return err
+	}
+
+	if err := are.fromArgoCDWrite.Flush(); err != nil {
+		logCtx.WithError(err).Info("unable to flush to Argo CD Redis socket")
+		return err
+	}
+
+	return nil
+}
+
+// establishConnectionToPrincipalRedis establishes a simple TCP-IP socket connection to principal's redis. (That is, we don't use go-redis client)
+func establishConnectionToPrincipalRedis(principalRedisAddress string, logCtx *logrus.Entry) (*net.TCPConn, error) {
+
+	var redisConn *net.TCPConn
+
+	addr, err := net.ResolveTCPAddr("tcp", principalRedisAddress)
+	if err != nil {
+		logCtx.WithError(err).WithField("redisAddress", principalRedisAddress).Error("Resolution error")
+		return nil, fmt.Errorf("unable to resolve address: %w", err)
+	}
+
+	// Dial the resolved address
+	redisConn, err = net.DialTCP("tcp", nil, addr)
+	if err != nil {
+		logCtx.WithError(err).WithField("redisAddress", principalRedisAddress).Error("Connection error")
+		return nil, fmt.Errorf("unable to connect to redis '%s': %w", principalRedisAddress, err)
+	}
+
+	return redisConn, nil
+}
+
+// Extract agent name from the key field of 'get' or 'subscribe' redis commands
+// If found, return name. If not found, return "" (with nil err).
+// If a field had an unexpected format, and error is returned.
+//
+// For example:
+// - If the key is 'app|managed-resources|agent-managed_my-app|1.8.3'
+// - The extracted agent name will be 'agent-managed'
+//   - This value comes from the third field, before the '_' character
+func extractAgentNameFromRedisCommandKey(redisKey string, logCtx *logrus.Entry) (string, error) {
+
+	// We only recognize agent names from these keys
+	if !strings.HasPrefix(redisKey, "app|managed-resources") && !strings.HasPrefix(redisKey, "app|resources-tree") {
+
+		if redisKey == "new-revoked-token" {
+			// 'new-revoked-token' is forwarded to principal
+			return "", nil
+		}
+
+		if strings.HasPrefix(redisKey, "mfst|") {
+			// mfst| is forwarded to principal, but I think this may not be a permanent behaviour.
+			// Example key:
+			// - mfst|annotation:app.kubernetes.io/instance|agent-managed_my-app|(revision id)|guestbook|3093507789|1.8.3.gz
+			logCtx.Info("redirecting mfst| redis key to principal redis")
+			return "", nil
+		}
```

Comment:
There is at least one more expected key name that should be redirected to the principal's Redis:

* `cluster|` (all related to cluster configuration)